{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import random\n",
    "import re\n",
    "import time\n",
    "import tqdm\n",
    "from tqdm import tqdm \n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "from torch.utils.data import TensorDataset, ConcatDataset, DataLoader\n",
    "from torch.utils.data import RandomSampler, SequentialSampler\n",
    "from transformers import AdamW, get_linear_schedule_with_warmup\n",
    "from transformers import AutoTokenizer, AutoModel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def set_seed(seed_value=42):\n",
    "    random.seed(seed_value)\n",
    "    np.random.seed(seed_value)\n",
    "    \n",
    "set_seed()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time taken: 27.3759\n"
     ]
    }
   ],
   "source": [
    "def load_data(filepath):\n",
    "    d = {}\n",
    "    with open(filepath) as f:\n",
    "        for i, line in enumerate(f):\n",
    "            d[i] = json.loads(line)\n",
    "    return pd.DataFrame.from_dict(d).T\n",
    "\n",
    "s = time.time()\n",
    "fullbody = load_data(\"/home/jessica/data/s2orc/s2orc_fullbody_subset_100K.jsonl\")\n",
    "print(f\"Time taken: {round(time.time()-s,4)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>abstract</th>\n",
       "      <th>year</th>\n",
       "      <th>arxiv_id</th>\n",
       "      <th>acl_id</th>\n",
       "      <th>pmc_id</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>doi</th>\n",
       "      <th>...</th>\n",
       "      <th>inbound_citations</th>\n",
       "      <th>has_outbound_citations</th>\n",
       "      <th>has_inbound_citations</th>\n",
       "      <th>has_pdf_parse</th>\n",
       "      <th>has_pdf_parsed_abstract</th>\n",
       "      <th>has_pdf_parsed_body_text</th>\n",
       "      <th>has_pdf_parsed_bib_entries</th>\n",
       "      <th>has_pdf_parsed_ref_entries</th>\n",
       "      <th>s2_url</th>\n",
       "      <th>body_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>94551546</td>\n",
       "      <td>Gd(III) ion-chelated supramolecular assemblies...</td>\n",
       "      <td>[{'first': 'Yu', 'middle': [], 'last': 'Zhao',...</td>\n",
       "      <td>An intricate polymer complex can carry genes t...</td>\n",
       "      <td>2015</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10.1038/am.2015.67</td>\n",
       "      <td>...</td>\n",
       "      <td>[543091]</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://api.semanticscholar.org/CorpusID:94551546</td>\n",
       "      <td>[{'section': 'INTRODUCTION', 'text': 'Gene the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18980380</td>\n",
       "      <td>Distributionally Robust Counterpart in Markov ...</td>\n",
       "      <td>[{'first': 'Pengqian', 'middle': [], 'last': '...</td>\n",
       "      <td>This technical note studies Markov decision pr...</td>\n",
       "      <td>2016</td>\n",
       "      <td>1501.07418</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10.1109/TAC.2015.2495174</td>\n",
       "      <td>...</td>\n",
       "      <td>[18980380, 202130625, 12287826, 195316976, 206...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://api.semanticscholar.org/CorpusID:18980380</td>\n",
       "      <td>[{'section': '', 'text': '. Illustration of th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18980463</td>\n",
       "      <td>Adult and larval photoreceptors use different ...</td>\n",
       "      <td>[{'first': 'S. G.', 'middle': [], 'last': 'Spr...</td>\n",
       "      <td>Although development of the adult Drosophila c...</td>\n",
       "      <td>2007</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PMC1950857</td>\n",
       "      <td>17785526</td>\n",
       "      <td>10.1101/gad.1565407</td>\n",
       "      <td>...</td>\n",
       "      <td>[201966468, 16579577, 14667576, 20024434, 1344...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://api.semanticscholar.org/CorpusID:18980463</td>\n",
       "      <td>[{'section': '', 'text': 'In spite of the morp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18981111</td>\n",
       "      <td>Exploration of Unknown Spaces by People Who Ar...</td>\n",
       "      <td>[{'first': 'Orly', 'middle': [], 'last': 'Laha...</td>\n",
       "      <td>Exploration of unknown spaces is essential for...</td>\n",
       "      <td>2004</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>10.1177/016264340401900302</td>\n",
       "      <td>...</td>\n",
       "      <td>[6869367, 24873259, 69173440, 12309066, 357763...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://api.semanticscholar.org/CorpusID:18981111</td>\n",
       "      <td>[{'section': '', 'text': 'The ability to explo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18981358</td>\n",
       "      <td>Rubber hand illusion induced by touching the f...</td>\n",
       "      <td>[{'first': 'Michele', 'middle': [], 'last': 'S...</td>\n",
       "      <td>Background: Studies in animals and humans indi...</td>\n",
       "      <td>2014</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>PMC4050649</td>\n",
       "      <td>24959128</td>\n",
       "      <td>10.3389/fnhum.2014.00404</td>\n",
       "      <td>...</td>\n",
       "      <td>[209325151, 206095004, 55701872, 5015858, 5802...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>https://api.semanticscholar.org/CorpusID:18981358</td>\n",
       "      <td>[{'section': 'INTRODUCTION', 'text': 'Spinal c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   paper_id                                              title  \\\n",
       "0  94551546  Gd(III) ion-chelated supramolecular assemblies...   \n",
       "1  18980380  Distributionally Robust Counterpart in Markov ...   \n",
       "2  18980463  Adult and larval photoreceptors use different ...   \n",
       "3  18981111  Exploration of Unknown Spaces by People Who Ar...   \n",
       "4  18981358  Rubber hand illusion induced by touching the f...   \n",
       "\n",
       "                                             authors  \\\n",
       "0  [{'first': 'Yu', 'middle': [], 'last': 'Zhao',...   \n",
       "1  [{'first': 'Pengqian', 'middle': [], 'last': '...   \n",
       "2  [{'first': 'S. G.', 'middle': [], 'last': 'Spr...   \n",
       "3  [{'first': 'Orly', 'middle': [], 'last': 'Laha...   \n",
       "4  [{'first': 'Michele', 'middle': [], 'last': 'S...   \n",
       "\n",
       "                                            abstract  year    arxiv_id acl_id  \\\n",
       "0  An intricate polymer complex can carry genes t...  2015        None   None   \n",
       "1  This technical note studies Markov decision pr...  2016  1501.07418   None   \n",
       "2  Although development of the adult Drosophila c...  2007        None   None   \n",
       "3  Exploration of unknown spaces is essential for...  2004        None   None   \n",
       "4  Background: Studies in animals and humans indi...  2014        None   None   \n",
       "\n",
       "       pmc_id pubmed_id                         doi  ...  \\\n",
       "0        None      None          10.1038/am.2015.67  ...   \n",
       "1        None      None    10.1109/TAC.2015.2495174  ...   \n",
       "2  PMC1950857  17785526         10.1101/gad.1565407  ...   \n",
       "3        None      None  10.1177/016264340401900302  ...   \n",
       "4  PMC4050649  24959128    10.3389/fnhum.2014.00404  ...   \n",
       "\n",
       "                                   inbound_citations has_outbound_citations  \\\n",
       "0                                           [543091]                   True   \n",
       "1  [18980380, 202130625, 12287826, 195316976, 206...                   True   \n",
       "2  [201966468, 16579577, 14667576, 20024434, 1344...                   True   \n",
       "3  [6869367, 24873259, 69173440, 12309066, 357763...                   True   \n",
       "4  [209325151, 206095004, 55701872, 5015858, 5802...                   True   \n",
       "\n",
       "  has_inbound_citations has_pdf_parse has_pdf_parsed_abstract  \\\n",
       "0                  True          True                    True   \n",
       "1                  True          True                    True   \n",
       "2                  True          True                    True   \n",
       "3                  True          True                   False   \n",
       "4                  True          True                    True   \n",
       "\n",
       "  has_pdf_parsed_body_text has_pdf_parsed_bib_entries  \\\n",
       "0                     True                       True   \n",
       "1                     True                       True   \n",
       "2                     True                       True   \n",
       "3                     True                       True   \n",
       "4                     True                       True   \n",
       "\n",
       "  has_pdf_parsed_ref_entries  \\\n",
       "0                       True   \n",
       "1                       True   \n",
       "2                       True   \n",
       "3                       True   \n",
       "4                       True   \n",
       "\n",
       "                                              s2_url  \\\n",
       "0  https://api.semanticscholar.org/CorpusID:94551546   \n",
       "1  https://api.semanticscholar.org/CorpusID:18980380   \n",
       "2  https://api.semanticscholar.org/CorpusID:18980463   \n",
       "3  https://api.semanticscholar.org/CorpusID:18981111   \n",
       "4  https://api.semanticscholar.org/CorpusID:18981358   \n",
       "\n",
       "                                           body_text  \n",
       "0  [{'section': 'INTRODUCTION', 'text': 'Gene the...  \n",
       "1  [{'section': '', 'text': '. Illustration of th...  \n",
       "2  [{'section': '', 'text': 'In spite of the morp...  \n",
       "3  [{'section': '', 'text': 'The ability to explo...  \n",
       "4  [{'section': 'INTRODUCTION', 'text': 'Spinal c...  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullbody.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['paper_id', 'title', 'authors', 'abstract', 'year', 'arxiv_id',\n",
       "       'acl_id', 'pmc_id', 'pubmed_id', 'doi', 'venue', 'journal',\n",
       "       'has_pdf_body_text', 'mag_id', 'mag_field_of_study',\n",
       "       'outbound_citations', 'inbound_citations', 'has_outbound_citations',\n",
       "       'has_inbound_citations', 'has_pdf_parse', 'has_pdf_parsed_abstract',\n",
       "       'has_pdf_parsed_body_text', 'has_pdf_parsed_bib_entries',\n",
       "       'has_pdf_parsed_ref_entries', 's2_url', 'body_text'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullbody.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paper_id</th>\n",
       "      <th>body_text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>94551546</td>\n",
       "      <td>[{'section': 'INTRODUCTION', 'text': 'Gene the...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18980380</td>\n",
       "      <td>[{'section': '', 'text': '. Illustration of th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18980463</td>\n",
       "      <td>[{'section': '', 'text': 'In spite of the morp...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18981111</td>\n",
       "      <td>[{'section': '', 'text': 'The ability to explo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18981358</td>\n",
       "      <td>[{'section': 'INTRODUCTION', 'text': 'Spinal c...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   paper_id                                          body_text\n",
       "0  94551546  [{'section': 'INTRODUCTION', 'text': 'Gene the...\n",
       "1  18980380  [{'section': '', 'text': '. Illustration of th...\n",
       "2  18980463  [{'section': '', 'text': 'In spite of the morp...\n",
       "3  18981111  [{'section': '', 'text': 'The ability to explo...\n",
       "4  18981358  [{'section': 'INTRODUCTION', 'text': 'Spinal c..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fullbody = fullbody[[\"paper_id\", \"body_text\"]]\n",
    "fullbody.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "94551546\n",
      "Gene therapy holds potential for treating many severe diseases, such as cancer and genetic diseases. 1 Successful gene therapy depends on highefficiency gene delivery processes, in which the gene carriers have an essential role. The application of traditional viral vectors has been a challenge because of their toxicity, immunogenicity and low capability for scaling up. 2 There has long been a scientific demand for developing non-viral gene delivery systems that can overcome the drawbacks of viral vectors. 3 Non-viral gene delivery has been advanced by the rapid development of materials science and technology. Numerous novel gene delivery systems have been proposed based on functional cationic polymers, such as polyethylenimine (PEI), [4] [5] [6] [7] [8] [9] poly(2-(dimethylamino) ethyl methacrylate), 4,10 poly(L-lysine), 11 poly(aspartic acid) 12, 13 and polyamidoamine. 14 However, these non-viral gene carriers still have shortcomings, including cytotoxicity, low transfection efficiency and lack of multifunction. Recently, we found that ethanolamine (EA) or ethylenediamine (ED)-functionalized poly(glycidyl methacrylate) (PGMA), namely PGEA or PGED, could be used as effective gene carriers. 15, 16 They possess good gene transfection properties. To further improve the performance of PGMA-based gene carriers, several strategies have been applied such as polysaccharide introduction and target molecule binding. 16, 17 Owing to the dynamically unable ability of supramolecular polymers, the application of supramolecular chemistry for gene delivery has been a hot research topic in the biomedical field. 18, 19 The construction of supramolecular polycations via host-guest interaction is a popular strategy for high-efficiency gene delivery systems. 20 In particular, cyclodextrins (CDs) and their derivatives have been widely utilized for constructing supramolecular gene delivery systems, mainly because of their superior biocompatibility. [21] [22] [23] With the host-guest interaction strategy, we successfully prepared one PGEA-based supramolecular delivery system by tying multiple β-cyclodextrin (β-CD)cored star PGEA polymers to an adamantine-modified linear PGEA backbone. 24 Such PGEA supramolecules markedly increased transfection efficiencies. Further improvements in functionality and the development of new preparation strategies for PGMA-based supramolecular vectors would benefit the construction of better gene delivery systems. To construct multifunctional supramolecular vectors, a novel strategy was proposed to flexibly prepare PGMA-based supramolecular delivery systems (PGED-Gd@PGEAs) with magnetic resonance imaging (MRI) functionality, by assembling multiple adamantine-headed star PGEA (Ad-PGEA) units with a versatile PGED-CD-Gd backbone (Scheme 1). This backbone possessed numerous flanking β-CD species and Gd 3+ ions. MRI technology has received considerable attention because of its high spatial resolution and its applications in areas such as deep tissue imaging. 25, 26 In particular, Gd 3+ ions have been used as contrast agents because of their optimal chemical and magnetic properties. 27, 28 In this work, Gd ions were chelated by diethylenetriaminepentacetate acid (DTPA) immobilized on the PGED backbones to introduce the MRI effect into the resultant PGED-Gd@PGEA supramolecular systems. The physicochemical properties of the PGED-Gd@PGEA assemblies, including plasmid DNA (pDNA) condensation ability, cytotoxicity, gene transfection, cellular uptake and MRI functionality, were examined in detail. The present work provides a new strategy to design multifunctional supramolecular delivery systems. Branched PEI (Mw~25 kDa), β-CD (99%), epichlorohydrin (EP, 99%), 1-adamantanecarboxylic acid chloride (98%), pentaerythritol (98%), 2-bromoisobutyryl bromide (98%), ethyl bromoisobutyrate (99%), glycidyl methacrylate (GMA, 98%), N,N,N′,N″,N″-pentamethyldiethylenetriamine (99%), copper(I) bromide (99%), EA (98%), ED (98%), DTPA (98%), carbodiimide hydrochloride (98%), gadolinium(III) chloride hexahydrate (99%) and 3-(4,5-dimethylthiazol-2yl)-2,5-diphenyl tetrazolium bromide were bought from Sigma-Aldrich Chemical Co., St Louis, MO, USA. GMA was used after removal of the inhibitors. The plasmid pRL-CMV, encoding Renilla luciferase, (Promega Co., Cergy Pontoise, France), and the plasmid pEGFP-N1, encoding enhanced green fluorescent protein (EGFP) (BD Biosciences, San      Linear PGMA was prepared via atom transfer radical polymerization (ATRP). First, 160 μl of ethyl bromoisobutyrate (1 equiv), 2.1 ml of GMA (60 equiv) and 300 μl of N,N,N′,N″,N″-pentamethyldiethylenetriamine (2 equiv) were added to a 50 ml flask with 5 ml of dimethyl sulfoxide (DMSO). The reaction system was degassed with argon for 10 min before adding 115 mg of copper(I) bromide (1 equiv). The details of the preparation of PGMA are described in our earlier work. 29 The molar weight of PGMA was 4.6 × 10 3 g mol −1 polydispersity index (PDI) = 1.28. The resultant ED-functionalized PGMA (PGED) was prepared using excess ED, as reported previously. 16 As shown in Scheme 1, before the preparation of CD units containing PGED (PGED-CD), it was necessary to synthesize EP-modified CD (CD-EP), using similar procedures to those reported previously. 30 In brief, 1.5 g of β-CD was added to a 50 ml round flask with a mixture (v/v, 5/5 ml) of DMSO/i-PrOH. After the β-CD was thoroughly dissolved, 12.5 ml of 1 M NaOH aqueous solution was then added to the solution, followed by the addition of 2 ml of EP. The reaction proceeded at room temperature in a nitrogen atmosphere for 48 h. The pH of the reaction solution was adjusted to~7.0 with concentrated HCl. Excess acetone was used to precipitate the reaction mixture. The raw CD-EP was dissolved in a small amount of deionized (DI) water and dialyzed against DI water (4 × 5 l) with a dialysis membrane (MWCO, 1000 Da) at room temperature for 4 h, prior to lyophilization. PGED-CD was prepared subsequently. In a 50 ml flask, PGED (130 mg) and CD-EP (480 mg) were added with 7 ml of water. The molar ratio of the terminal amino groups (of PGED) and CD-EP units was 1.6:1. The reaction was conducted at 80°C for 48 h. The PGED-CD product was purified by a dialysis method and then lyophilized. For the preparation of Gd 3+ -chelated PGED-CD (PGED-CD-Gd), the amidation reaction was used to introduce DTPA as the chelating agent of Gd 3+ ions onto PGED-CD according to procedures described previously. 31 In brief, 8 mg of carbodiimide hydrochloride and 40 mg of DTPA were dissolved in 7 ml of water. The molar ratio of carbodiimide hydrochloride and DTPA was kept at 1:5 to avoid cross-linking between the amino groups of PGED-CD and carboxylate groups of DTPA. The solution was stirred for 4 h at room temperature before adding 100 mg of PGED-CD. The molar ratio of the remaining amino groups (of PGED-CD) and DTPA units was 1:1. The reaction proceeded for 48 h to produce PGED-CD-DTPA. The resulting solution was purified by dialyzed with a dialysis membrane (MWCO, 1000 Da) against 0.1 M NaCl for 2 days and then dialyzed against DI water for 1 day. The purified solution of PGED-CD-DTPA was transferred into a flask, and GdCl 3 (GdCl 3 •6H 2 O) was added. The molar ratio of DTPA and GdCl 3 (GdCl 3 •6H 2 O) was kept at 1:1. The mixing solution was stirred for 24 h at room temperature and then dialyzed against DI water for 24 h. Subsequently, the PGED-CD-Gd product was lyophilized. For the preparation of the Ad-PGEA guest, the adamantine-headed ATRP initiator (Ad-Br) with three initiation sites was first synthesized. Pentaerythritol (1.08 g, 7.94 mmol) was thoroughly dissolved in a 50 ml round flask containing 20 ml of anhydrous N,N-dimethylformamide. Then, 1-adamantanecarboxylic acid chloride (1 g, 5.05 mmol) and K 2 CO 3 (1.33 g, 9.64 mmol) were added. The reaction was conducted at 50°C for 24 h under magnetic stirring. The final reaction solution was centrifuged, evaporated and distillated under reduced pressure, producing Ad-OH with three hydroxyl groups. The resultant Ad-Br was prepared using the similar procedures described earlier. 29 Next, 447 mg of Ad-OH (1.5 mmol) was added to a 50 ml round flask containing 7 ml of N,Ndimethylformamide. Then, 0.75 ml of 2-bromoisobutyryl bromide (6 mmol) was dropped into the aforementioned solution under an ice bath condition and stirring for 24 h. The reaction mixture was quenched with water and extracted with CH 2 Cl 2 . The organic layer was washed with brine, dried over Na 2 SO 4 and concentrated under reduced pressure. The Ad-Br product was a yellowish syrupy-like liquid. The resultant Ad-PGMA polymers were synthesized under the typical conditions of ATRP. 29 In a 50 ml flask, 120 mg of Ad-Br (1 equiv), 2.3 ml of GMA (100 equiv) and 70 μl of N,N,N′,N″,N″-pentamethyldiethylenetriamine (2.5 equiv) were added to a 50 ml round flask containing 5 ml of DMSO. The reaction system was degassed by nitrogen for 10 min before adding 22.8 mg copper(I) bromide (1 equiv) under a nitrogen atmosphere. The products with a polymerization time of 20 and 35 min were named Ad-PGMA1 (Mn = 5.5 × 10 3 g mol −1 , PDI = 1.26) and Ad-PGMA2 (Mn = 1.0 × 10 4 g mol −1 , PDI = 1.32), respectively. The resulting EA-functionalized Ad-PGMA (Ad-PGEA) was prepared using excess EA as reported earlier. 15 The crude product was purified using a dialysis membrane (MWCO 3500), followed by lyophilization. For the preparation of PGED@PGEA and PGED-Gd@PGEA assemblies, with a 1:1 molar feed ratio of the CD/Ad units, PGED-CD or PGED-CD-Gd was assembled with Ad-PGEA. The concentration of Ad-PGEA was based on the nitrogen concentration of 20 mM. Before use, equal volumes of Ad-PGEA solution and PGED-CD or PGED-CD-Gd solution were mixed, shaken and incubated at ambient temperature for 3 h. The final nitrogen concentration for all polymer solutions was 10 mM in DI water. The polymer to DNA ratio was expressed as the molar ratio of nitrogen (N) in PGMA-based polycations to phosphate (P) in DNA, named as the N/P ratio. The average mass weight per phosphate group of DNA was assumed to be 325. 32 All PGED@PGEA/pDNA and PGED-Gd@PGEA/pDNA polyplexes at different N/P ratios were formed by mixing polymer solution and pDNA solution for 30 min before use. 1 H NMR spectra were measured on a Bruker ARX 400 MHz spectrometer using CDCl 3 (for Ad-Br and Ad-PGMA) and D 2 O (for Ad-PGEA, PGED, CD-EP, PGED-CD, PGED-CD-DTPA) as the solvents with tetramethylsilane (Me 4 Si) as an internal standard. GPC measurements of PGMA and Ad-PGMA were performed on a Waters GPC system, in which DMSO was used as the eluent at a low flow rate of 1.0 ml min −1 at 25°C and monodispersed poly (methyl methacrylate) standards were used to generate the calibration curve. Dynamic light scattering measurements of polyplexes were performed with a Zetasizer Nano ZS (Malvern Instruments, Southborough, MA, USA) equipped with a laser of wavelength 633 nm at a 173°scattering angle. Atomic force microscopy (AFM) studies were carried out with the Dimension Icon model with a Nanoscope IIIa controller (Bruker, Santa Barbara, CA, USA). The samples were imaged using the ScanAsyst mode. Image analysis was performed using Nanoscope software after removing the background slope after flattening the images. Gel electrophoresis was implemented in a Sub-Cell system (Bio-Rad Laboratories), and then, a UV transilluminator and BioDco-It imaging system (UVP Inc.) was used to record DNA bands. C6 and Hep G2 cell lines from two common cancers were selected to evaluate the performance of the gene carriers. An 3-(4,5-dimethylthiazol-2yl)-2,5diphenyl tetrazolium bromide assay was used to evaluate the cytotoxicity of polyplexes at a series of N/P ratios in C6 and Hep G2 cells cultured in DMEM (with 10% fetal bovine serum) according to the methods described in our previous studies. 32, 33 In brief, the C6 and Hep G2 cells were cultured in 96-well plates at a density of 2 × 10 4 cells per well with DMEM (with 10% fetal bovine serum). Then, the culture medium was replaced with 100 μl of fresh culture medium. The PGEA-based and PEI complexes (6.7 μl per well containing 0.33 μg of pDNA) at various N/P ratios were added to the media. Then, the cells were incubated for 4 h, and fresh media were added to the culture for another 20 h. Then, 10 μl of sterile, filtered 3-(4,5-dimethylthiazol-2yl)-2,5diphenyl tetrazolium bromide solution in phosphate-buffered saline (PBS) (5 mg ml −1 ) was added to each well. After 4 h, the unreacted dye was removed, and the produced formazan crystals were dissolved in DMSO (100 μl per well). The OD values were measured at a wavelength of 570 nm with a Bio-Rad Model 680 Microplate Reader (UK). For each sample, the final absorbance was the average of those measured from six wells in parallel. The cell viability results were expressed as the percentage relative to that of the control. The plasmid pRL-CMV as the reporter gene was first utilized to estimate the in vitro gene transfection of PGMA-based polycation/pDNA polyplexes in C6 and Hep G2 cell lines. Essentially, the C6 and Hep G2 cells were cultured in 24well plates at a density of 6 × 10 4 cells per well with DMEM media (with 10% fetal bovine serum). Then, the solutions of polycation/pDNA complexes with different N/P ratios (20 μl per well with 1.0 μg of pDNA) were added into the transfection media. The detailed transfection procedures are described in our earlier work. 15, 16, 32, 33 A commercial Promega kit and a luminometer (Berthold Lumat LB 9507, Berthold Technologies GmbH KG, Bad Wildbad, Germany) were used to quantify the luciferase gene expression. Gene expression results were expressed as relative light units per milligram of cell protein lysate (relative light units per mg protein). The plasmid pEGFP-N1 with EGFP gene was also utilized as the reporter gene in C6 and Hep G2 cell lines at the optimal N/P ratio of the polycations to evaluate polymer-mediated gene transfection. The transfected cells were imaged with a Leica DMI3000B fluorescence microscope. The percentage of the EGFP-positive cells was determined by flow cytometry (Beckman Coulter, Brea, CA, USA). The in vitro cellular uptake was determined by flow cytometry analysis and imaged using fluorescence microscopy. C6 and Hep G2 cells were seeded into 24-well plates at the density of 8 × 10 5 cells per well and incubated for 24 h. Then, the cells were incubated with the fresh media containing PGMA-based polycation/pDNA polyplexes for 4 h, where pDNA (pRL-CMV) were labeled with the fluorescent dye YOYO-1. 34 The cells were trypsinized, centrifuged, resuspended in PBS and then analyzed by flow cytometry (BD LSR II, BD, USA). For the fluorescence imaging, after cellular uptake, the cells were washed with PBS five times and stained with 4′,6-diamidino-2-phenylindole for 10 min. The fluorescence images were acquired on a Leica DMI3000B microscope. Approximately 5 × 10 6 C6 or Hep G2 cells were seeded and incubated in cell culture flasks for 24 h. Then, the media were replaced with 5 ml of fresh media containing PGED-CD-Gd or PGED-Gd@PGEA ((Gd) = 0, 5, 10 and 20 μM, respectively). After incubation at 37°C for 4 h, the cells were washed with PBS three times and then precipitated by centrifugation. The MRI experiments were performed on a 7.0-T MRI instrument (BioSpec 70/20 USR 7.0 T Bruker) with a wrist receiver coil. The pulse sequence was a T1 map-RATE sequence with the following parameters: TR/TE = 400, 800, 1500, 2500, 4000/7 ms: field of view: 3.5 cm 2 ; matrix: 128 × 128; number of excitations: 1; slice thickness = 1 mm; slice gap = 0 mm; field of view: 3.5 cm. PGED-Gd@PGEA or detached cells were prepared in PBS at different concentrations in 0.5 ml Eppendorf tubes. The T1 relaxivities were determined via a linear fit of the inverse relaxation time as a function of the Gd 3+ concentration. All experiments were repeated at least three times. The data are presented as means ± s.d. Statistical significance (Po0.05) was evaluated by a t-test when two groups of samples were compared, and the statistical significance was set at Po0.05. Preparation and characterization of supramolecular polycations with Gd 3+ chelation As shown in Scheme 1, the PGMA-based supramolecular delivery systems (PGED-Gd@PGEAs) with MRI functions were prepared by assembling multiple Ad-PGEA units with a PGED-CD-Gd backbone. The PGED-CD-Gd backbone possessed numerous flanking β-CD species and Gd 3+ ions. For the preparation of the PGED-CD-Gd backbone, the starting PGMA (Mn = 4.6 × 10 3 g mol −1 , PDI= 1.28) with 32 repeat units was first synthesized via ATRP. Then, PGMA was functionalized with excess ED by ring-opening addition to produce PGED with abundant primary amine groups. Similar to our previous study, 16 it was confirmed that the PGED was successfully prepared from the NMR results ( Supplementary Figure S1(a1) ). To introduce β-CD units, CD-EP that possessed one epoxy ring (Supplementary Figure S1(a2) ) was immobilized onto PGED via ring-opening Figure S1(a3) ), the molar ratio of β-CD and ED units was 1:2, indicating that every PGED-CD chain contained 16 β-CD units. DTPA, the chelating agent of Gd 3+ ions, was subsequently reacted with the remaining primary amine groups of PGED-CD via the amidation reaction, producing PGED-CD-DTPA. Based on the 1 H NMR spectrum of PGED-CD-DTPA ( Supplementary  Figure S1(a4) ), PGED-CD-DTPA contained~12 DPTA units. The resultant PGED-CD-Gd backbone was obtained by the chelation of PGED-CD-DTPA with Gd 3+ ions. According to the thermogravimetric analysis (Supplementary Figure S2) , the PGED-CD-Gd backbone contained~10 Gd 3+ ions, which was fairly consistent with the number of DPTA units. The synthesis of the Ad-PGEA guest with three arms was performed in a four-step procedure (Scheme 1). C(CH 2 OH) 4 was compounded with 1-adamantanecarboxylic acid chloride, and the rest of the hydroxyl groups were reacted with 2-bromoisobutyryl bromide to initiate the ATRP process of GMA, followed by ring-opening of epoxy groups of Ad-PGMA with excess EA. Every step of the preparation process of Ad-PGEA was monitored by 1 H NMR (Supplementary Figure S3 ). After the first two steps of the substitution reaction, C-CH 2 -OCO protons and CH 3 -C-Br protons from the newly formed Ad-OH and Ad-Br were observed at~3.89 and 1.95 ppm, respectively. By integrating the corresponding NMR peaks, it was estimated that Ad-Br possessed approximately three initiation sites. The resultant Ad-PGMA1 (Mn = 5.5 × 10 3 g mol −1 , PDI= 1.26) and Ad-PGMA2 (Mn = 1.0 × 10 4 g mol −1 , PDI = 1.32) were prepared by using Ad-Br. The Ad-PGMA species were further functionalized with excess EA to produce the corresponding Ad-PGEA1 and Ad-PGEA2 guests with different molecular weights. Similar to our previous studies, 15, 16 Ad-PGMA and Ad-PGEA were successfully prepared, as shown by the NMR results (Supplementary Figure S3 ). For the preparation of the PGED@PGEA and PGED-Gd@PGEA assemblies, two Ad-PGEA guests were complexed with PGED-CD or PGED-CD-Gd hosts in DI water by the host-guest interaction of Ad and CD species. Characterization of polymer/pDNA nanocomplexes For non-viral polycationic gene carriers, the ability to condense pDNA into a nanoparticle is essential because nanocomplexes of carrier and pDNA are more suitable for cellular uptake. In this work, agarose gel electrophoresis, particle size and ζ-potential measurements, as well as AFM images, were used to confirm the ability of PGMA-based polycations to condense pDNA. The gel retardation results of polymer/pDNA complexes with increasing N/P ratios are shown in Figure 1 . All PGED@PGEA and PGED-Gd@PGEA could compact pDNA completely within the N/P ratio of 1.5, whereas the PGED-CD-Gd backbone and Ad-PGEA guests only condensed pDNA when N/P ratios reached 2.5. The above results indicated that the pDNAcondensing ability of PGED@PGEA and PGED-Gd@PGEA is better than that of PGED-CD-Gd and Ad-PGEA, which is probably because of the high molecular weight supramolecular structures. 35 The particle size and the ζ-potential of polymer/pDNA complexes with increasing N/P ratio are shown in Figure 2 . The particle sizes of all complexes decreased with increasing N/P ratios. At lower N/P ratios, loose aggregates with large size were formed owing to the few positive charges on the polymers. 36 With an increase in the N/P ratios, the diameters of all complexes decreased to 100-150 nm, which enabled the complexes to easily undergo endocytosis. It was also noted that at both low and high N/P ratios, the polycation/pDNA complexes had good stability (Supplementary Figure S4 ). In addition to particle sizes, the ζ-potential of polymer/pDNA complexes are another factor that affects cellular uptake. All of the complexes had positive ζ-potentials, and the values of the ζ-potential slightly increased with the N/P ratios. Owing to the positive charges, polymer/pDNA complexes had a good affinity for the negatively charged cell membrane and were able to facilitate cellular uptake. 37 The ζ-potential values of PGED@PGEA and PGED-Gd@PGEA groups were higher than those of the PGED-CD-Gd backbone and Ad-PGEA guests, once again confirming that PGMA-based supramolecular polycations possessed better DNA-compacting ability. The morphology of polymer/pDNA complexes was observed by AFM. Figure 3 shows the representative AFM images of Ad-PGEA1/ pDNA, PGED@PGEA1/pDNA and PGED-Gd@PGEA1/pDNA complexes at the N/P ratio of 15. The AFM images show that the polymers could compact pDNA to form nanoparticles with a diameter of 100 nm, which is consistent with the results of particle size measurement (Figure 2a ). Low cytotoxicity is essential for ideal biomedical materials. To evaluate the cytotoxicity of PGMA-based polycation/pDNA complexes at different N/P ratios, an 3-(4,5-dimethylthiazol-2yl)-2,5-diphenyl tetrazolium bromide assay was performed using the C6 and Hep G2 cell lines. The relative cell viabilities of all of the groups in both cell lines decreased with increasing N/P ratios (Figure 4 ). The cytotoxicity of polycations generally increases with their molecular weight. 38 In comparison with the PGED-CD and PGED-CD-Gd backbone, PGED@PGEA and PGED-Gd@PGEA exhibited increased cytotoxicity. However, PGED@PGEA and PGED-Gd@PGEA showed significantly lower cytotoxicity than the gold standard, PEI (25 kDa). For example, at an N/P ratio of 30, for the PGMA-based polycation/pDNA complexes, the relative cell viability values of C6 cells were 450% and those of Hep G2 470%. In our earlier work, 3, 15, 24, 39, 40 it was found that in comparison with PEI, which mainly has secondary amine groups, the nonionic hydrophilic hydroxyl groups of PGEAs can shield the deleterious, excess charges of the cationic complexes, leading to lower cytotoxicity. In addition, no significant differences among the cytotoxicities of Ad-PGEA, PGED@PGEA and PGED-Gd@PGEA were observed. The above results indicated that the supramolecular polymers could achieve a lower cytotoxicity despite their high molecular weights. The gene transfection efficiency of the polycation/pDNA complexes was first analyzed in C6 and Hep G2 cell lines using pRL-CMV as a reporter gene. The C6 and Hep G2 cell lines are documented to be difficult to transfect with synthetic vectors. 41 Figure 5 shows the gene transfection efficiencies mediated by Ad-PGEA, PGED@PGEA and PGED-Gd@PGEA at various N/P ratios in comparison with those mediated by the controls (PEI (25 kDa), PGED-CD and PGED-CD-Gd) at their respective optimal N/P ratios. In general, the transfection efficiency first increases and then decreases slightly with increasing N/P ratio. At lower N/P ratios, cationic polymers cannot efficiently compact DNA, and the resultant loose nanocomplexes cannot easily enter cells. By contrast at higher N/P ratios, the transfection efficiency slightly decreased, probably as a result of the increased toxicity of free cationic polymers. The difference in gene transfection efficiencies at the same N/P ratio was observed between C6 and Hep G2 cells. The transfection efficiencies of gene carriers depend on different cell types because of their different properties. 42 As expected, the transfection efficiencies of PGED@PGEA and PGED-Gd@PGEA were significantly higher than those of Ad-PGEA guests at various N/P ratios because PGED@PGEA and  In vitro gene transfection efficiencies of the polymer/pDNA complexes at various N/P ratios in (a) C6 and (b) Hep G2 cells in comparison with those mediated by PEI (25 kDa, at its optimal N/P ratio of 10), PGED-CD (at the optimal N/P ratio of 15 for C6 or 25 for G2 cells) and PGED-CD-Gd (at the optimal N/P ratio of 15 for C6 or 25 for Hep G2 cells) (mean ± s.d., n = 3, *Po0.05). PGED-Gd@PGEA might benefit from the better DNA-compacting ability (Figure 1 ) and low toxicities (Figure 4 ) of the supramolecular vectors. Notably, the transfection efficiencies of PGED@PGEA and PGED-Gd@PGEA arising from the same Ad-PGEA guests exhibited no significant difference, indicating that the Gd 3+ ions had no obvious effects on transfection efficiencies. In C6 and Hep G2 cell lines, Ad-PGEA2 exhibited much higher transfection efficiencies than Ad-PGEA1. This indicated that the transfection efficiencies mediated by PGEA-based vectors were dependent on the molecular weight of PGEA. In addition, the transfection efficiencies mediated by PGED@P-GEA and PGED-Gd@PGEA at most N/P ratios were much higher than those mediated by PEI at its optimal ratio of 10. As shown in Figure 5 , the differences between the transfection efficiencies of Ad-PGEA1 and PGED-Gd@PGEA1 at most N/P ratios are bigger than those of Ad-PGEA2 and PGED-Gd@PGEA2. Thus, PGED-Gd@PGEA1 was selected as the representative supramolecular vector in the following experiments. To visually confirm the gene delivery abilities of PGEA-based vectors, the plasmid pEGFP-N1, Figure 6 Representative images of EGFP expression in (a, a1, a2, a3) C6 and (b, b1, b2, b3) Hep G2 cells transfected with Ad-PGEA1, PGED@PGEA1, PGED-Gd@PGEA1 and PEI at their respective optimal N/P ratios. Biomedical applications of polycations Y Zhao et al Figure 7 Fluorescent images and flow cytometry analysis plots of (a) C6 and (b) Hep G2 cells treated with Ad-PGEA1/pDNA, PGED@PGEA1/pDNA and PGED-Gd@PGEA1/pDNA polyplexes for 4 h at the respective optimal N/P ratios. For the fluorescent images, the YOYO-1-labeled pDNA is shown in green, and the 4′,6-diamidino-2-phenylindole -labeled nuclei are shown in blue. Y Zhao et al encoding GFP, was used as another reporter gene in C6 and Hep G2 cell lines. Representative images of EGFP gene expression mediated by Ad-PGEA1, PGED@PGEA1 and PGED-Gd@PGEA1 at their respective optimal N/P ratios are shown in Figure 6 . The corresponding merged images are shown in Supplementary Figure S5 . Transfection with PGED@PGEA1 or PGED-Gd@PGEA1 led to more green (EGFPpositive) Hep G2 cells than did Ad-PGEA1 or PEI. The transfection efficiency, as reflected by the percentage of EGFP-positive cells, was quantitatively determined using flow cytometry. The percentages of EGFP-positive cells for Ad-PGEA1, PGED@PGEA1, PGED-Gd@PGEA1 and PEI in C6 (or Hep G2) cells are 13% (or 16%), 43% (or 47%), 41% (or 44%) and 14% (or 23%), respectively, which is fairly consistent with the results for luciferase expression ( Figure 5 ). The above gene transfection assay indicated that the supramolecular structure improved the gene transfection, and Gd 3+ ions had no obvious adverse effects on the transfection process. To investigate the intracellular fates of the supramolecular vectors, the C6 and Hep G2 cells were treated with the Ad-PGEA1/pDNA, PGED@PGEA1/pDNA and PGED-Gd@PGEA1/pDNA complexes for 4 h at their respective optimal N/P ratios (15 for C6 cells and 25 for Hep G2 cells), whereas the pDNA was labeled by YOYO1, as shown in Figure 7 . YOYO1-labeled pDNA is shown in green. The nuclei stained with 4′,6-diamidino-2-phenylindole are shown in blue. Compared with the cells treated with Ad-PGEA1/pDNA, the cells treated with PGED@PGEA1/pDNA and PGED-Gd@PGEA1/pDNA showed more green aggregations. The corresponding merged images are shown in Supplementary Figure S6 . The cellular uptake of the complexes with YOYO1-labeled pDNA was quantified by flow cytometry. The percentages of YOYO1-positive cells treated with PGED@PGEA1/ pDNA and PGED-Gd@PGEA1/pDNA were almost 90%, much higher than those (65~69%) of the cells treated with Ad-PGEA1/pDNA. This result was consistent with the fluorescent images of cellular uptake, confirming that supramolecular structure enhanced cell uptake, benefiting the resultant gene transfection. The complexes with gadolinium have been widely used as efficient T1weighted MR contrast agents. 41, 43 PGED-Gd@PGEA was expected to act as an MRI contrast agent (relaxation times were shown in Supplementary Table S1 ). The MRI assay of PGED-Gd@PGEA1 is shown in Figure 8 . A linear curve could be obtained by plotting the inverse T1 as a function of the molar concentration of Gd 3+ ions (Figure 8a, top) . The brightness of the MR images increased with the molar concentration of Gd 3+ ions (Figure 8a, bottom) . Next, the potential to use the PGED-Gd@PGEA1 for MRI of cancer cells was explored. Both C6 and Hep G2 cells were treated for 4 h with PGED-CD-Gd and PGED-Gd@PGEA1 at different Gd 3+ ion concentrations in culture media (0, 5, 10 and 20 μM,) before imaging with the MR systems. Both the inverse T1 and MR images were presented and collected. The MR images of cancer cells treated with PGED-CD-Gd became slightly brighter with an increasing concentration of Gd 3+ ions (Supplementary Figure S7) . However, the MR images of cancer cells treated with PGED-Gd@PGEA1 evidently became brighter, and the inverse T1 increased with the concentration of Gd 3+ ions (Figure 8b ), likely arising from the higher cellular uptake of PGED-Gd@PGEA1 (Figure 7) . The above results indicated that supramolecular structure could also affect the relaxation time in vitro. PGED-Gd@PGEA could potentially be used as a contrast agent for MR imaging. A series of PGMA-based supramolecular polycations (PGED-Gd@PGEAs) with MRI functionality were successfully constructed by assembling multiple Ad-PGEA guests with a versatile PGED-CD-Gd host, which possessed numerous flanking Gd 3+ ions. Such PGED-Gd@PGEAs had good DNA condensation abilities and low cell cytotoxicity. Moreover, PGED-Gd@PGEAs exhibited significantly higher transfection efficiencies than PEI (25 kDa) and the constituent units (PGED-CD-Gd and Ad-PGEA). The chelation of Gd 3+ ions imparted PGED-Gd@PGEAs with effective MRI functionality without adverse effects on gene transfection processes. These unique features could allow PGED-Gd@PGEA to become a competitive multifunctional gene delivery system. Figure 8 Linear fitting of the inverse T 1 and the T 1 -weighted MR images of (a) PGED-Gd@PGEA1 solutions and (b) C6 and Hep G2 cells treated with PGED-Gd@PGEA1 at different Gd concentrations.\n"
     ]
    }
   ],
   "source": [
    "# work with one paper for now\n",
    "paper = fullbody.loc[0]\n",
    "print(paper[\"paper_id\"])\n",
    "print(\" \".join([_[\"text\"] for _ in paper[\"body_text\"]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialise sentence tokenizer\n",
    "from nltk.tokenize.punkt import PunktSentenceTokenizer, PunktParameters\n",
    "\n",
    "punkt_param = PunktParameters()\n",
    "\n",
    "# tokenizer should not split at abbrieviations\n",
    "punkt_param.abbrev_types = set([\"i.e\", \"etc\", \"al\", \"fig\", \"figs\", \"ref\", \"refs\", \"p\", \"c\", \"s\"]) \n",
    "punkt_sent_tokenizer = PunktSentenceTokenizer(punkt_param)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Recently, we found that ethanolamine (EA) or ethylenediamine (ED)-functionalized poly(glycidyl methacrylate) (PGMA), namely PGEA or PGED, could be used as effective gene carriers. 15, 16 They possess good gene transfection properties. To further improve the performance of PGMA-based gene carriers, several strategies have been applied such as polysaccharide introduction and target molecule binding. 16, 17 Owing to the dynamically unable ability of supramolecular polymers, the application of supramolecular chemistry for gene delivery has been a hot research topic in the biomedical field. 18, 19 The construction of supramolecular polycations via host-guest interaction is a popular strategy for high-efficiency gene delivery systems. 20 In particular, cyclodextrins (CDs) and their derivatives have been widely utilized for constructing supramolecular gene delivery systems, mainly because of their superior biocompatibility. [21] [22] [23] With the host-guest interaction strategy, we successfully prepared one PGEA-based supramolecular delivery system by tying multiple β-cyclodextrin (β-CD)cored star PGEA polymers to an adamantine-modified linear PGEA backbone. 24 Such PGEA supramolecules markedly increased transfection efficiencies. Further improvements in functionality and the development of new preparation strategies for PGMA-based supramolecular vectors would benefit the construction of better gene delivery systems.\n",
      "\n",
      "[22]: [21] [22] [23] With the host-guest interaction strategy, we successfully prepared one PGEA-based supramolecular delivery system by tying multiple β-cyclodextrin (β-CD)cored star PGEA polymers to an adamantine-modified linear PGEA backbone.\n",
      "[23]: [21] [22] [23] With the host-guest interaction strategy, we successfully prepared one PGEA-based supramolecular delivery system by tying multiple β-cyclodextrin (β-CD)cored star PGEA polymers to an adamantine-modified linear PGEA backbone.\n"
     ]
    }
   ],
   "source": [
    "# (try to) extract citation sentence\n",
    "citation_sentences = []\n",
    "for paragraph in paper[\"body_text\"]:\n",
    "    cite_spans = paragraph[\"cite_spans\"]\n",
    "    \n",
    "    # ignore paragraphs without citations\n",
    "    if not cite_spans: continue \n",
    "        \n",
    "    paragraph_text = paragraph[\"text\"]\n",
    "    print(paragraph_text)\n",
    "    print()\n",
    "        \n",
    "    # tokenize the paragraph into sentences\n",
    "    endpoints = list(punkt_sent_tokenizer.span_tokenize(paragraph_text))\n",
    "    \n",
    "    # for each citation marker, \n",
    "    for cite_span in cite_spans:\n",
    "        start, end = cite_span[\"start\"], cite_span[\"end\"]\n",
    "        \n",
    "        # print the sentence containing the citation marker\n",
    "        for a,b in endpoints:\n",
    "            if start >= a and end <= b:\n",
    "                print(f\"{cite_span['text']}: {paragraph_text[a:b]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Recently, we found that ethanolamine (EA) or ethylenediamine (ED)-functionalized poly(glycidyl methacrylate) (PGMA), namely PGEA or PGED, could be used as effective gene carriers. 15, 16 They possess good gene transfection properties. To further improve the performance of PGMA-based gene carriers, several strategies have been applied such as polysaccharide introduction and target molecule binding. 16, 17 Owing to the dynamically unable ability of supramolecular polymers, the application of supramolecular chemistry for gene delivery has been a hot research topic in the biomedical field. 18, 19 The construction of supramolecular polycations via host-guest interaction is a popular strategy for high-efficiency gene delivery systems. 20 In particular, cyclodextrins (CDs) and their derivatives have been widely utilized for constructing supramolecular gene delivery systems, mainly because of their superior biocompatibility. [21] [22] [23] With the host-guest interaction strategy, we successfully prepared one PGEA-based supramolecular delivery system by tying multiple β-cyclodextrin (β-CD)cored star PGEA polymers to an adamantine-modified linear PGEA backbone. 24 Such PGEA supramolecules markedly increased transfection efficiencies. Further improvements in functionality and the development of new preparation strategies for PGMA-based supramolecular vectors would benefit the construction of better gene delivery systems.\n",
      "\n",
      "[22]: 20 In particular, cyclodextrins (CDs) and their derivatives have been widely utilized for constructing supramolecular gene delivery systems, mainly because of their superior biocompatibility. [21] [22]\n",
      "[23]: 20 In particular, cyclodextrins (CDs) and their derivatives have been widely utilized for constructing supramolecular gene delivery systems, mainly because of their superior biocompatibility. [21] [22] [23]\n"
     ]
    }
   ],
   "source": [
    "# use regex to extract citation sentences correctly\n",
    "citation_sentences = []\n",
    "for paragraph in paper[\"body_text\"]:\n",
    "    cite_spans = paragraph[\"cite_spans\"]\n",
    "    \n",
    "    # ignore paragraphs without citations\n",
    "    if not cite_spans: continue \n",
    "        \n",
    "    paragraph_text = paragraph[\"text\"]\n",
    "    print(paragraph_text)\n",
    "    print()\n",
    "        \n",
    "    # tokenize the paragraph into sentences\n",
    "    endpoints = list(punkt_sent_tokenizer.span_tokenize(paragraph_text))\n",
    "    \n",
    "    # for each citation marker, \n",
    "    for cite_span in cite_spans:\n",
    "        cite_text = cite_span[\"text\"]\n",
    "        start, end = cite_span[\"start\"], cite_span[\"end\"]\n",
    "        \n",
    "        # determine if citation marker is textual\n",
    "        textual = re.search('[a-zA-Z]', cite_text) \n",
    "        \n",
    "        # find the sentence containing the citation marker\n",
    "        for i, pair in enumerate(endpoints):\n",
    "            a,b = pair\n",
    "            if start >= a and end <= b: \n",
    "                candidate = paragraph_text[a:b]\n",
    "                \n",
    "                # if citation marker is textual or candidate begins with words,  \n",
    "                if textual or re.search('[a-zAZ]', paragraph_text[a:start]): \n",
    "                    # assume candidate is correct citation sentence\n",
    "                    citation_sentences.append((cite_text, candidate))\n",
    "                else:\n",
    "                    # previous sentence is true citation sentence \n",
    "                    a,b = endpoints[i-1]\n",
    "                    citation_sentences.append((cite_text, paragraph_text[a:end]))\n",
    "\n",
    "for pair in citation_sentences:\n",
    "    print(f\"{pair[0]}: {pair[1]}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The generation of the two types of ommatidia, yellow and pale, includes several steps. First, the stochastic expression of the transcription factor Spineless (Ss) in a subset of R7 cells specifies yellow ommatidia. Ss is required cell autonomously in yR7 for rh4 expression and, further, cell nonautonomously for the underlying R8 cell to acquire y fate and turn on rh6 expression (Wernet et al. 2006) . The coordination between R7 and R8 rhodopsins requires a signal from pR7 that induces the pR8 fate. In sevenless mutants that lack R7, rh5 expression is lost while rh6 is expanded to almost all R8 (Papatsenko et al. 1997; Chou et al. 1999) . The y versus p choice in R8 is then reinforced by a bistable loop of regulation between the tumor suppressor gene warts (wts) and the growth regulator melted (melt) (Mikeladze-Dvali et al. 2005b) : wts is required for rh6 expression, whereas melt is essential for rh5 expression. wts and melt repress each other transcriptionally, thereby ensuring that a robust decision to express either rh5 or rh6 is made. The ho-meoprotein encoded by orthodenticle (otd) is required in both p R7 and R8 to activate rh3 and rh5 through direct binding to their promoter. Otd has only a permissive role: It is expressed in all PRs and its overexpression is not able to induce rh3 or rh5 expression in y ommatidia. Furthermore, the y ommatidial fate does not expand in otd mutants (Tahayato et al. 2003) .\n",
      "\n",
      "(Wernet et al. 2006): Ss is required cell autonomously in yR7 for rh4 expression and, further, cell nonautonomously for the underlying R8 cell to acquire y fate and turn on rh6 expression (Wernet et al. 2006) .\n",
      "(Mikeladze-Dvali et al. 2005b): The y versus p choice in R8 is then reinforced by a bistable loop of regulation between the tumor suppressor gene warts (wts) and the growth regulator melted (melt) (Mikeladze-Dvali et al. 2005b) : wts is required for rh6 expression, whereas melt is essential for rh5 expression.\n",
      "(Tahayato et al. 2003): Furthermore, the y ommatidial fate does not expand in otd mutants (Tahayato et al. 2003) .\n",
      "\n",
      "\n",
      "Although much is known about the development of the adult visual system, little is known about the development and function of the larval visual system. PRs of the larval eye (Bolwig Organ [BO] ) extend their axonal projections to the larval pacemaker neurons, which control the larval circadian rhythm (Malpel et al. 2002; Hassan et al. 2005; Moncalvo and Campos 2005) . Visual input via the larval eye is crucial for the entrainment of the molecular clock (Malpel et al. 2004; Mazzoni et al. 2005) . Furthermore, both larval PRs and pacemaker neurons control larval photophobic behavior (Mazzoni et al. 2005) . Very simple in structure, the larval eye contains ∼12-14 PRs (Green et al. 1993) . In contrast to adult ommatidia, BO lacks accessory cells, such as pigment cells or cone cells.\n",
      "\n",
      "Mazzoni et al. 2005): Visual input via the larval eye is crucial for the entrainment of the molecular clock (Malpel et al. 2004; Mazzoni et al. 2005) .\n",
      "(Mazzoni et al. 2005): Furthermore, both larval PRs and pacemaker neurons control larval photophobic behavior (Mazzoni et al. 2005) .\n",
      "\n",
      "\n",
      "The sal genes encode two zinc finger transcription factors that are specifically expressed in adult inner PRs R7 and R8, where they are required to distinguish them from outer PRs (Mollereau et al. 2001 ). Since sal is exclusively expressed in the Rh5 subtype, prior to rh5 expression, we tested whether it is required for the development of this subtype. We thus analyzed the expression of Rh5 and Rh6 in sal mutants at the end of embryogenesis, when these mutants die. No expression of Rh5 can be detected in sal mutants, even though the correct Cold (Fig. 6C) . However, the general PR markers Elav, FasII, Kr, and Chp remain expressed in all PRs, suggesting that four PRs express neither Rh5 nor Rh6 ( Fig. 6C ; data not shown). Since rh1, rh3, and rh4 expression cannot be detected, these four PRs appear to be \"empty\" and devoid of PR molecules (data not shown). PRs of the larval eye project into the region of the dendritic arborizations of the larval pacemaker neurons. Axon termini of Rh5 and Rh6 PRs are generally directly adjacent to each other (Mazzoni et al. 2005) . In sal mutants, all larval PRs project into the correct target region of the late embryonic brain (Fig. 6G) . We compared the projection termini of empty PRs with those of Rh6 PRs (\"empty\" PR termini are identified by FasII staining and the absence of Rh6 staining). At embryonic stage 17 (just before these mutants die), axonal termini of \"empty\" PRs project correctly to the target region, adjacent to Rh6 termini (Fig.  6H,I ). Therefore sal is essential in the larval eye for the proper differentiation of the Rh5 PRs but not for their early specification or for axonal targeting.\n",
      "\n",
      "(Mazzoni et al. 2005): Axon termini of Rh5 and Rh6 PRs are generally directly adjacent to each other (Mazzoni et al. 2005) .\n",
      "\n",
      "\n",
      "otd is required in the Rh5 subtype for rh5 expression and rh6 repression The homeodomain protein Otd is required in inner PRs of the adult retina for the activation of rh3 and rh5 in p ommatidia, whereas it is required in outer PRs to repress rh6 (Tahayato et al. 2003 ). In the adult retina, otd is expressed in all PRs. Similarly, otd is expressed in all embryonic immature PRs, and this expression is maintained during larval life. During early development, otd acts as a major component for patterning the anterior region of the embryo (Cohen and Jurgens 1990) . During stages 8/9, otd expression spans the entire cephalic region (Fig. 7A) and becomes subsequently restricted to more anterior regions and regions giving rise to the brain. However otd expression is excluded from large parts of the optic lobe primordium as well as from the region giving rise to larval PRs (Fig. 7B) . Starting at stage 12, however, otd expression is reinitiated in the ventral lateral part of the optic lobe primordium, the region that will give rise to the precursors of larval PRs. otd remains expressed in all PRs throughout embryogenesis and larval life (Fig. 7C-F) . otd is not required for the formation of larval PRs (nor the larval optic lobe primordium): Even though otd-null mutant embryos show severe head involution defects, the normal number of immature PRs is formed (data not shown). To investigate whether otd is involved in larval PR-subtype specification, we analyzed Rh5 and Rh6 expression in viable otd uvi mutants ( Vandendries et al. 1996) . In otd uvi , larval PRs do not express Rh5 while the total number of PRs remains the same (Fig. 7G,H ) and all PRs express Rh6 (Fig. 7G) . About four cells still express sal, but these cells now express Rh6 instead of Rh5 (Fig. 7H) . Interestingly, the number of Svp-expressing cells is not altered, with four Svp-negative PRs expressing Rh6 (data not shown). This indicates that, in the Rh5 subtype, otd acts as an inhibitor of rh6 expression and is required for Rh5 expression. As expression of otd is unaltered in sal mutants (data not shown), otd, like sal, seems to be necessary but not sufficient for Rh5 expression. otd and sal expression does not depend on each other, and since they are both required for Rh5 expression, they appear to act in parallel pathways.\n",
      "\n",
      "(Tahayato et al. 2003: otd is required in the Rh5 subtype for rh5 expression and rh6 repression The homeodomain protein Otd is required in inner PRs of the adult retina for the activation of rh3 and rh5 in p ommatidia, whereas it is required in outer PRs to repress rh6 (Tahayato et al. 2003 ).\n",
      "\n",
      "\n",
      "Specification of adult PRs starts with the proliferation of undifferentiated cells anterior to the morphogenetic furrow and the recruitment of individual PRs into ommatidia posterior to the morphogenetic furrow in a tightly regulated spatiotemporal manner. R8 is first specified by ato and does not depend on EGFR signaling (Jarman et al. 1994; Freeman 1996) . Sequential recruitment of all other PRs (R2/5, R3/4, R1/6, and then R7) is dependent on EGFR signaling (Freeman 1996; Raabe 2000; Nagaraj and Banerjee 2004) . Similarly, in the larval eye, primary precursors express ato and are independent of EGFR signaling while secondary precursors need EGFR signaling for their development. Primary precursors develop into the Rh5 subtype while the Rh6-subtype identity corresponds to secondary precursors.\n",
      "\n",
      "Freeman 1996): R8 is first specified by ato and does not depend on EGFR signaling (Jarman et al. 1994; Freeman 1996) .\n",
      "(Freeman 1996;: Sequential recruitment of all other PRs (R2/5, R3/4, R1/6, and then R7) is dependent on EGFR signaling (Freeman 1996; Raabe 2000; Nagaraj and Banerjee 2004) .\n",
      "Raabe 2000;: Sequential recruitment of all other PRs (R2/5, R3/4, R1/6, and then R7) is dependent on EGFR signaling (Freeman 1996; Raabe 2000; Nagaraj and Banerjee 2004) .\n",
      "\n",
      "\n",
      "During larval PR development, EGFR signaling is required for the Rh6 but not the Rh5 subtype. tll inhibits this process by preventing the adjacent optic lobe primordium from responding to EGFR signaling (Daniel et al. 1999; Dumstrei et al. 2002; Chang et al. 2003a ). However, expression of tll is not negatively regulated by EGFR signaling (Daniel et al. 1999; Dumstrei et al. 2002; Chang et al. 2003a) . Blocking apoptosis or removing tll function both result in supernumerary Rh6 cells, indicating that cells that should have died in response to tll function become Rh6 PRs (Fig. 8A) . Secondary precursors do not appear to require maintenance of EGFR signaling to survive. During later developmental stages, all immature PRs express sevenless, another gene encoding a Receptor Tyrosine Kinase as well as its ligand, Boss (S.G. Specher and C. Desplan, unpubl.), which could act redundantly with EGFR later in development. However, we see no effect of mutating boss or sev, either on larval PR number or on Rh5/Rh6 expression (S.G. Specher and C. Desplan, unpubl.). It could be that late EGFR activity compensates for the loss of Sev activity, comparable with the adult R7 cells, where EGFR is sufficient to replace Sev (Freeman 1996) .\n",
      "\n",
      "(Freeman 1996): It could be that late EGFR activity compensates for the loss of Sev activity, comparable with the adult R7 cells, where EGFR is sufficient to replace Sev (Freeman 1996) .\n",
      "\n",
      "\n",
      "During larval eye development, sal is only required for the expression of Rh5 but not for the specification of Rh5-subtype fate. First, ectopic activation of Sal is not sufficient to induce Rh5 expression. Further, sal is still expressed in cells that have adopted the Rh6 subtype due to ectopic svp expression. In contrast, svp is required and sufficient for the Rh6 subtype where it represses the Rh5-subtype fate. Interestingly, svp is not only required for the repression of sal, but is also necessary for Rh6 expression (Fig. 8A) . In the adult retina, svp is necessary for the specification of the R3/R4 and R1/R6 pairs (Mlodzik et al. 1990; Domingos et al. 2004a,b) where it is also required for the repression of sal: R3/R4 are transformed into R7 cells in svp mutants (Domingos et al. 2004b) . However, the upstream mechanisms by which the expression of sal and svp is controlled in larval PRs remains elusive. There may be an unknown signal from primary precursors that induces secondary precursors by controlling svp expression. This signal is probably not EGFR, since EGFR dn secondary cells that are rescued from death by p35 still express svp.\n",
      "\n",
      "(Domingos et al. 2004b): In the adult retina, svp is necessary for the specification of the R3/R4 and R1/R6 pairs (Mlodzik et al. 1990; Domingos et al. 2004a,b) where it is also required for the repression of sal: R3/R4 are transformed into R7 cells in svp mutants (Domingos et al. 2004b) .\n",
      "\n",
      "\n",
      "In the adult retina otd is required for the expression of rh3 and rh5 and for the repression of Rh6 in outer PRs. In the larval eye, is also expressed in all PRs, but it is not required for the formation of larval PRs. During their terminal differentiation, otd is required, only in the Rh5 subtype, for Rh5 expression and Rh6 repression (Fig. 8A) . Since otd only functions in the context of sal expression, it acts as a permissive factor for Rh regulation. It seems likely that otd and sal act in parallel in the Rh5 subtype: Otd expression is not altered in sal mutants and Sal expression is not altered in otd mutants. Further, Otd, which binds directly to the rh3, rh5, and rh6 enhancers in the adult eye, likely acts in a similar fashion in larval PRs (Tahayato et al. 2003) .\n",
      "\n",
      "(Tahayato et al. 2003): Further, Otd, which binds directly to the rh3, rh5, and rh6 enhancers in the adult eye, likely acts in a similar fashion in larval PRs (Tahayato et al. 2003) .\n",
      "\n",
      "\n",
      "For wild-type comparison we used yw 122 ; yw 122 , sp/CyO, TM2/ TM6b; or heterozygous siblings of mutant alleles. For sal mutant analysis, the sal 16 and a small Deficiency-Df(2L)32FP-5, which removes salm and salr-were used, balanced over CyO, Dfd-YFP; both fly strains gave comparable results (Kuhnlein et al. 1994; Elstob et al. 2001) . For svp mutant analysis, the svp E22 was used, balanced over TM6b, Dfd-YFP (Mlodzik et al. 1990 ). For tll mutant analysis, the tll 1 and tll 149 were used, balanced over TM6b, Dfd-YFP. For sal, tll, and svp, homozygous mutants were identified by the absence of Dfd-YFP (Bloomington Stock Center). We used the viable otd uvi allele (Vandendries et al. 1996) . The following fly strains were used: so-Gal4 (Chang et al. 2003b) , sev-Gal4 (Therrien et al. 1999) , otd-Gal4 (T. Cook, pers. comm.), ato-Gal4 (Hassan et al. 2000) , svp H162 -LacZ (Elstob et al. 2001 (Kuhnlein and Schuh 1996) , UASsvp (Kramer et al. 1995) , UAS-CD8GFP, UAS-p35, UAS-pnt\n",
      "\n",
      "(Therrien et al. 1999): The following fly strains were used: so-Gal4 (Chang et al. 2003b) , sev-Gal4 (Therrien et al. 1999) , otd-Gal4 (T. Cook, pers.\n",
      "\n",
      "\n",
      "Embryos were dechorionated, fixed, and immunostained according to previously published protocols (Therianos et al. 1995) . Primary antibodies were rabbit anti-Rh6 1:10,000 (Tahayato et al. 2003) , mouse anti-Rh5 1:20, anti-Rh3 1:20, or anti-Rh4 1:20 (Chou et al. 1996) , mouse anti-Nrt 1:10 (Developmental Studies Hybridoma Bank [DSHB]), mouse anti-Rh1 1:20 (DSHB), mouse anti-FasII 1:10 (Lin and Goodman 1994), rat anti-Elav 1:30 (DSHB), goat anti-Ato 1:1000 (Jarman et al. 1993) , sheep anti-GFP (Biogenesis), rabbit anti-Sal 1:200 (Kuhnlein et al. 1994) , mouse anti-Svp 1:1000 (Kanai et al. 2005) , mouse antiPros 1:50 (DSHB), mouse anti-Chp 1:10 (DSHB), rat anti-Otd 1:200 (Hirth et al. 2003) , rat anti-Kr 1:300 (Kosman et al. 1998) , rab anti-Pnt P1 (Alvarez et al. 2003) , anti-Yan (Rebay and Rubin 1995) , guinea pig anti-Ss 1:500 (Kim et al. 2006) , and mouse anti-␤GAL 1:20 (DSHB). Secondary antibodies used for confocal microscopic analysis were Alexa-488, Alexa-555, and Alexa-647 antibodies generated in goat (Molecular probes), all at 1:300-1:500 dilution. Embryos were mounted in Vectashield H-1000 (Vector). For the analysis of the larval BO, the head skeleton was separated from epidermis, imaginal discs, and brain and fixed for 15 min in 4% formaldehyde/PBS. The chitinous head skeleton was then carefully opened on the dorsal and ventral midline using sharpened minutien pins (0.1-mm diameter, Fisher Scientific Tools).\n",
      "\n",
      "(Tahayato et al. 2003): Primary antibodies were rabbit anti-Rh6 1:10,000 (Tahayato et al. 2003) , mouse anti-Rh5 1:20, anti-Rh3 1:20, or anti-Rh4 1:20 (Chou et al. 1996) , mouse anti-Nrt 1:10 (Developmental Studies Hybridoma Bank [DSHB]), mouse anti-Rh1 1:20 (DSHB), mouse anti-FasII 1:10 (Lin and Goodman 1994), rat anti-Elav 1:30 (DSHB), goat anti-Ato 1:1000 (Jarman et al. 1993) , sheep anti-GFP (Biogenesis), rabbit anti-Sal 1:200 (Kuhnlein et al. 1994) , mouse anti-Svp 1:1000 (Kanai et al. 2005) , mouse antiPros 1:50 (DSHB), mouse anti-Chp 1:10 (DSHB), rat anti-Otd 1:200 (Hirth et al. 2003) , rat anti-Kr 1:300 (Kosman et al. 1998) , rab anti-Pnt P1 (Alvarez et al. 2003) , anti-Yan (Rebay and Rubin 1995) , guinea pig anti-Ss 1:500 (Kim et al. 2006) , and mouse anti-␤GAL 1:20 (DSHB).\n",
      "(Hirth et al. 2003): Primary antibodies were rabbit anti-Rh6 1:10,000 (Tahayato et al. 2003) , mouse anti-Rh5 1:20, anti-Rh3 1:20, or anti-Rh4 1:20 (Chou et al. 1996) , mouse anti-Nrt 1:10 (Developmental Studies Hybridoma Bank [DSHB]), mouse anti-Rh1 1:20 (DSHB), mouse anti-FasII 1:10 (Lin and Goodman 1994), rat anti-Elav 1:30 (DSHB), goat anti-Ato 1:1000 (Jarman et al. 1993) , sheep anti-GFP (Biogenesis), rabbit anti-Sal 1:200 (Kuhnlein et al. 1994) , mouse anti-Svp 1:1000 (Kanai et al. 2005) , mouse antiPros 1:50 (DSHB), mouse anti-Chp 1:10 (DSHB), rat anti-Otd 1:200 (Hirth et al. 2003) , rat anti-Kr 1:300 (Kosman et al. 1998) , rab anti-Pnt P1 (Alvarez et al. 2003) , anti-Yan (Rebay and Rubin 1995) , guinea pig anti-Ss 1:500 (Kim et al. 2006) , and mouse anti-␤GAL 1:20 (DSHB).\n",
      "(Alvarez et al. 2003): Primary antibodies were rabbit anti-Rh6 1:10,000 (Tahayato et al. 2003) , mouse anti-Rh5 1:20, anti-Rh3 1:20, or anti-Rh4 1:20 (Chou et al. 1996) , mouse anti-Nrt 1:10 (Developmental Studies Hybridoma Bank [DSHB]), mouse anti-Rh1 1:20 (DSHB), mouse anti-FasII 1:10 (Lin and Goodman 1994), rat anti-Elav 1:30 (DSHB), goat anti-Ato 1:1000 (Jarman et al. 1993) , sheep anti-GFP (Biogenesis), rabbit anti-Sal 1:200 (Kuhnlein et al. 1994) , mouse anti-Svp 1:1000 (Kanai et al. 2005) , mouse antiPros 1:50 (DSHB), mouse anti-Chp 1:10 (DSHB), rat anti-Otd 1:200 (Hirth et al. 2003) , rat anti-Kr 1:300 (Kosman et al. 1998) , rab anti-Pnt P1 (Alvarez et al. 2003) , anti-Yan (Rebay and Rubin 1995) , guinea pig anti-Ss 1:500 (Kim et al. 2006) , and mouse anti-␤GAL 1:20 (DSHB).\n",
      "(Rebay and Rubin 1995): Primary antibodies were rabbit anti-Rh6 1:10,000 (Tahayato et al. 2003) , mouse anti-Rh5 1:20, anti-Rh3 1:20, or anti-Rh4 1:20 (Chou et al. 1996) , mouse anti-Nrt 1:10 (Developmental Studies Hybridoma Bank [DSHB]), mouse anti-Rh1 1:20 (DSHB), mouse anti-FasII 1:10 (Lin and Goodman 1994), rat anti-Elav 1:30 (DSHB), goat anti-Ato 1:1000 (Jarman et al. 1993) , sheep anti-GFP (Biogenesis), rabbit anti-Sal 1:200 (Kuhnlein et al. 1994) , mouse anti-Svp 1:1000 (Kanai et al. 2005) , mouse antiPros 1:50 (DSHB), mouse anti-Chp 1:10 (DSHB), rat anti-Otd 1:200 (Hirth et al. 2003) , rat anti-Kr 1:300 (Kosman et al. 1998) , rab anti-Pnt P1 (Alvarez et al. 2003) , anti-Yan (Rebay and Rubin 1995) , guinea pig anti-Ss 1:500 (Kim et al. 2006) , and mouse anti-␤GAL 1:20 (DSHB).\n",
      "\n",
      "\n",
      "For laser confocal microscopy, a Leica TCS SP was used. Optical sections ranged from 0.2 to 1.5 µm, recorded in line average mode with a picture size of 512 × 512 pixels or 1024 × 1024 pixels. Captured images from optical sections were arranged and processed using Leica confocal Software (LCS). Complete series of optical sections were imported and processed using ImageJ. Generation of three-dimensional digital models and raw tiff stacks (stacks of optical sections) were done using AMIRA (Mercury Computer Systems) as previously described (Sprecher et al. 2006) .\n",
      "\n",
      "(Sprecher et al. 2006): Generation of three-dimensional digital models and raw tiff stacks (stacks of optical sections) were done using AMIRA (Mercury Computer Systems) as previously described (Sprecher et al. 2006) .\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# try the above on papers with different citation marker types \n",
    "paper = fullbody.loc[2]\n",
    "\n",
    "citation_sentences = []\n",
    "for paragraph in paper[\"body_text\"]:\n",
    "    cite_spans = paragraph[\"cite_spans\"]\n",
    "    if not cite_spans: continue\n",
    "    \n",
    "    paragraph_text = paragraph[\"text\"]\n",
    "    print(paragraph_text)\n",
    "    print()\n",
    "    \n",
    "    endpoints = list(punkt_sent_tokenizer.span_tokenize(paragraph_text))\n",
    "\n",
    "    for cite_span in cite_spans:\n",
    "        cite_text = cite_span[\"text\"]\n",
    "        start, end = cite_span[\"start\"], cite_span[\"end\"]\n",
    "        textual = re.search('[a-zA-Z]', cite_text) \n",
    "\n",
    "        for i, pair in enumerate(endpoints):\n",
    "            a,b = pair\n",
    "            if start >= a and end <= b: \n",
    "                candidate = paragraph_text[a:b]\n",
    "                if textual or re.search('[a-zAZ]', paragraph_text[a:start]): \n",
    "                    citation_sentence = candidate\n",
    "                else:\n",
    "                    a,b = endpoints[i-1]\n",
    "                    citation_sentence = paragraph_text[a:end]\n",
    "\n",
    "        print(f\"{cite_text}: {citation_sentence}\")\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# deploy on sample set\n",
    "extracted = {}\n",
    "\n",
    "s = time.time()\n",
    "for i in tqdm(range(len(fullbody))):\n",
    "    paper_id = fullbody.loc[i][\"paper_id\"]\n",
    "    bodytext = fullbody.loc[i][\"body_text\"]\n",
    "    \n",
    "    citation_sentences = []\n",
    "    for paragraph in bodytext:\n",
    "        cite_spans = paragraph[\"cite_spans\"]\n",
    "        if not cite_spans: continue\n",
    "\n",
    "        paragraph_text = paragraph[\"text\"]\n",
    "\n",
    "        endpoints = list(punkt_sent_tokenizer.span_tokenize(paragraph_text))\n",
    "\n",
    "        for cite_span in cite_spans:\n",
    "            cite_text = cite_span[\"text\"]\n",
    "            start, end = cite_span[\"start\"], cite_span[\"end\"]\n",
    "            textual = re.search('[a-zA-Z]', cite_text) \n",
    "\n",
    "            for i, pair in enumerate(endpoints):\n",
    "                a,b = pair\n",
    "                if start >= a and end <= b: \n",
    "                    candidate = paragraph_text[a:b]\n",
    "                    if textual or re.search('[a-zAZ]', paragraph_text[a:start]): \n",
    "                        citation_sentences.append((cite_text, candidate))\n",
    "                    else:\n",
    "                        a,b = endpoints[i-1]\n",
    "                        citation_sentences.append((cite_text, paragraph_text[a:end]))\n",
    "    \n",
    "    if citation_sentences:\n",
    "        extracted[paper_id] = citation_sentences\n",
    "print(f\"Time taken: {round(time.time()-s,4)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print(len(extracted))\n",
    "print(random.choice(list(extracted.values())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
